src/main/scala/com/fulcrumgenomics/sv/tools/AggregateSvPileup.scala

```diff
@@ -0,0 +1,319 @@
+package com.fulcrumgenomics.sv.tools
+
+import com.fulcrumgenomics.FgBioDef.{FgBioEnum, FilePath}
+import com.fulcrumgenomics.sopt.{arg, clp}
+import com.fulcrumgenomics.sv.BreakpointPileup
+import com.fulcrumgenomics.sv.cmdline.{ClpGroups, SvTool}
+import com.fulcrumgenomics.sv.tools.AggregatedBreakpointPileup.addId
+import com.fulcrumgenomics.sv.tools.BreakpointCategory.BreakpointCategory
+import com.fulcrumgenomics.util.{Io, Metric}
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+
+
+/** Type aliases */
+trait AggregateSvPileupDef {
+  type PileupGroup = Seq[BreakpointPileup]
+}
+
+
+@clp(group=ClpGroups.All, description=
+  """
+    |Merges nearby pileups of reads supporting putative structural variants.
+    |
+    |Takes as input the file of pileups produced by `SvPileup`. That file contains a list of breakpoints, each
+    |consisting of a chromosome, position and strand for each side of the breakpoint, as well as quantified read support
+    |for the breakpoint.
+    |
+    |This tool merges pairs of breakpoints that have their left and right ends on the same chromosomes, the same left
+    |and right strands, and their left and right positions both within a length threshold. The merging behavior is
+    |transitive. For example, two breakpoints that are farther apart than the length threshold can be merged if there
+    |is a third breakpoint that is close to both.
+    |
+    |The output file is a tab-delimited table with one record per aggregated cluster of pileups. Aggregated
+    |pileups are reported with the minimum and maximum (inclusive) coordinates of all pileups in the cluster, a
+    |possible putative structural variant event type supported by the pileups, and the sum of read support from all
+    |pileups in the cluster.
+    |""")
+class AggregateSvPileup
+(@arg(flag='i', doc="The input text file of pileups generated by SvPileup") input: FilePath,
+ @arg(flag='o', doc="The output file") output: FilePath,
+ @arg(flag='d', doc="Distance threshold below which to cluster breakpoints") maxDist: Int = 100,
+) extends SvTool with AggregateSvPileupDef {
+
+  Io.assertReadable(input)
+  Io.assertCanWriteFile(output)
+
+  override def execute(): Unit = {
+    // Read pileups from input file
+    val pileups: Seq[BreakpointPileup] = Metric.read[BreakpointPileup](input)
+
+    // Group pileups by left and right contig and strand
+    val pileupsByContigStrand: Map[BreakpointContigOrientation, PileupGroup] =
+      pileups.groupMap(BreakpointContigOrientation(_))(identity)
+
+    // Within each contig+strand group, identify all immediate "neighbors" of each pileup (other pileups within the
+    // distance threshold)
+    val pileupToNeighbors: Map[BreakpointContigOrientation, Map[BreakpointPileup, PileupGroup]] =
+      pileupsByContigStrand.view.mapValues(AggregateSvPileup.pileupToNeighbors(_, maxDist)).toMap
```

Comment:
Why not just sort the pileup group by `left_pos` then `right_pos`, then it's easy to linearly iterate to do:

```scala
if (prev.left_pos - cur.left_pos <= maxDist && prev.right_pos - cur.right_pos <= maxDist) {
  // add to current group
} else {
 // finalize previous group, and add the cur to a new one
}
```

what am I missing?

Reply:
@nh13 here's a counterexample:

|id|left_pos|right_pos|
|---|---|---|
|1|1000|5000|
|2|1000|7000|
|3|1001|5000|
|4|1001|7000|

These will sort in the order listed: `1`, `2`, `3`, `4`. You will get no clusters by the method you described. However, there should be two clusters: `1,3` and `2,4`.

Reply:
You're right!